xla/service/gpu/runtime/nccl_all_to_all_thunk.h

```diff
@@ -74,10 +74,11 @@ class NcclAllToAllStartThunk : public NcclCollectiveThunk {
   const std::vector<Buffer> buffers_;
   int64_t device_count_ = 1;
   bool p2p_memcpy_enabled_ = false;
+  absl::Mutex send_mutex_, receive_mutex_;
   absl::node_hash_map<int64_t, absl::node_hash_map<int64_t, uint64_t>>
-      send_pointer_maps_;
+      send_pointer_maps_ ABSL_GUARDED_BY(send_mutex_);
```

Comment:
`ABSL_GUARDED_BY` does not actually protect anything. It's simply an annotation that can be used by a clang analyzer to detect unprotected access.

Now that you have the mutex you need to lock it by creating an `absl::MutexLock lock(&send_mutex_)`  object in the right places.

Reply:
thanks! updated

Reply:
Thanks.

I see that you added locking in a few places, but here are more locations in the `.cc` file that are currently unprotected. You need to protect all accesses (e.g. `RunNcclCollective` contains additional accesses).

Also, please confirm that you have run the failing test **on 2 GPUs** a large number of times to verify that there is no deadlocks or races. You can use `blaze test ... --runs_per_test=1000`

Reply:
There's an unprotected access to send and receive map each, but these two are only sharing the piece with index of rank, i.e. only the same rank can ever access the same piece. Given this is a distributed cache, we cannot lock the whole map since different ranks need to access it at the same time. We can add locks for each of these pieces but that would be redundant since no lock will ever be shared across different ranks.

I've updated the granularity of the locks and verified that  with `--runs_per_test=1000` the test works fine now. Thanks!